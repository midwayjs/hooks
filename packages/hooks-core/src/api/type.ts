import type { Merge } from 'type-fest'
import { AsyncFunction } from '../types/common'

export type ArrayToObject<T, R = {}> = T extends [infer First, ...infer Rest]
  ? First extends PromiseLike<infer PromiseValue>
    ? PromiseValue
    : First extends object
    ? Merge<First, ArrayToObject<Rest, R>>
    : ArrayToObject<Rest, R>
  : R

export type ApiRunner<
  Input extends object | void,
  Handler extends AsyncFunction
> = (
  ...args: Input extends void
    ? Parameters<Handler>
    : [...args: Parameters<Handler>, input: Input]
) => ReturnType<Handler>

export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}

export type MetadataHelper = {
  setMetadata: <T = any>(key: any, value: T) => void
  getMetadata: <T = any>(key: any) => T
}

export type ExecuteHelper = {
  result?: any
  getInputArguments?: () => any[]
}

export type Operator<Input> = {
  name: string
  type?: Input
  input?: boolean
  metadata?: (helper: MetadataHelper) => void
  execute?: (helper: ExecuteHelper, next: () => Promise<any>) => Promise<void>
}

export type ExtractInputType<T> = {
  [key in keyof T]: T[key] extends Operator<any> ? T[key]['type'] : void
}

export type BaseTrigger = {
  type: string
  requestClient?: {
    fetcher: string
    client: string
  }
  [key: string]: any
}
