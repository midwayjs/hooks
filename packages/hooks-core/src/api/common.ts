import { isFunction } from '../common'

export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}

export const isApiFunction = (target: any): boolean => {
  return (
    isFunction(target) && !!Reflect.getMetadata(OperatorType.Trigger, target)
  )
}
