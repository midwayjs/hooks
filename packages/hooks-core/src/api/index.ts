import { BaseTrigger, OperatorType } from './type'

export function getApiTrigger<T extends BaseTrigger = BaseTrigger>(
  target: any
): T {
  return Reflect.getOwnMetadata(OperatorType.Trigger, target)
}

export * from './api'
export * from './type'

export * from './operator/middleware'
export * from './operator/http'
export * from './operator/validate'
