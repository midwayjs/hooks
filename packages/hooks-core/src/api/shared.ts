export enum OperatorType {
  Trigger = 'Trigger',
  Middleware = 'Middleware',
}
