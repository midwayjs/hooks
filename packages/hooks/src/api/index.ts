export * from './hooks'
export * from './hoc'
export * from './configuration'
export { HooksComponent as hooks } from './component'
export {
  Api,
  All,
  Get,
  Post,
  Put,
  Delete,
  Patch,
  Head,
  Options,
  Query,
  Params,
  Header,
  Middleware,
  Redirect,
  HttpCode,
  SetHeader,
  ContentType,
} from '@midwayjs/hooks-core'
export * from './operator/validate'
export * from './operator/serverless'
export * from './operator/type'
