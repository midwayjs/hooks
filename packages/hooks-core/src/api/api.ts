import 'reflect-metadata'
import { AsyncFunction, validateFunction } from '../'
import { AbstractFrameworkAdapter } from '../adapter/framework'
import { USE_INPUT_METADATA } from '../common/const'
import compose from 'koa-compose'
import { HttpMetadata } from './operator/http'
import {
  ApiRunner,
  ArrayToObject,
  ExecuteHelper,
  ExtractInputType,
  MetadataHelper,
  Operator,
} from './type'

export function Api<
  Operators extends Operator<any>[],
  Handler extends AsyncFunction
>(
  ...args: [...operators: Operators, handler: Handler]
): ApiRunner<
  ExtractInputType<Operators> extends void[]
    ? void
    : ArrayToObject<ExtractInputType<Operators>>,
  Handler
> {
  const handler = args.pop() as Function
  validateFunction(handler, 'ApiHandler')

  const operators = args as Operator<any>[]

  const metadataHelper: MetadataHelper = {
    getMetadata(key: any) {
      return Reflect.getMetadata(key, runner)
    },
    setMetadata(key: any, value: any) {
      return Reflect.defineMetadata(key, value, runner)
    },
  }

  for (const operator of operators) {
    if (operator.metadata) {
      validateFunction(operator.metadata, 'operator.metadata')
      operator.metadata(metadataHelper)
    }
  }

  const useInputMetadata = operators.some((operator) => operator.input)
  const executors = operators
    .filter((operator) => typeof operator.execute === 'function')
    .map((operator) => operator.execute)

  async function runner(...args: any[]) {
    const stack = [...executors]

    const executeHelper: ExecuteHelper = {
      result: undefined,
      getInputArguments: () => args,
    }

    stack.push(async (helper: ExecuteHelper, next: any) => {
      helper.result = await handler(...args)
      return next()
    })

    await compose(stack)(executeHelper)

    // handle HttpCode/Redirect/etc.
    const responseMetadata = Reflect.getMetadata(HttpMetadata.RESPONSE, runner)
    if (Array.isArray(responseMetadata)) {
      await AbstractFrameworkAdapter.instance.handleResponseMetaData(
        responseMetadata
      )
    }
    return executeHelper.result
  }

  Reflect.defineMetadata(USE_INPUT_METADATA, useInputMetadata, runner)
  return runner as any
}
