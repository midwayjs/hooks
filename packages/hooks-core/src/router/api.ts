import { basename, extname, removeExt } from 'upath'
import { AbstractRouter } from './base'
import urlJoin from 'proper-url-join'
import { isApiFunction } from '../api/common'
import { API_BASE_PATH } from '../common'
import isObject from 'lodash/isObject'

export type ApiRouterConfig = {
  basePath?: string
}

export class ApiRouter extends AbstractRouter {
  constructor(public config: ApiRouterConfig = {}) {
    super()
    config.basePath ??= API_BASE_PATH
  }

  isApiFile(options): boolean {
    return this.hasExportApiRoutes(options.mod)
  }

  hasExportApiRoutes(mod: object) {
    return isObject(mod) && Object.values(mod).some(isApiFunction)
  }

  getFunctionId(
    file: string,
    functionName: string,
    exportDefault: boolean
  ): string {
    return exportDefault
      ? removeExt(basename(file), extname(file))
      : functionName
  }

  functionToHttpPath(
    file: string,
    functionName: string,
    exportDefault: boolean
  ): string {
    return urlJoin(
      this.config.basePath,
      this.getFunctionId(file, functionName, exportDefault),
      { trailingSlash: false }
    )
  }
}
