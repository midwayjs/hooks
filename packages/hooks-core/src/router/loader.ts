import pickBy from 'lodash/pickBy'
import {
  ApiModule,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  FunctionId,
  HooksMiddleware,
} from '../'
import { USE_INPUT_METADATA } from '../common/const'
import { BaseTrigger, HttpTrigger, HttpTriggerType, OperatorType } from '../api'
import { ApiFunction } from '../types'
import { AbstractRouter } from './base'
import { createDebug, isFunction } from '../common'

const debug = createDebug('hooks-core: loader')

type Trigger = BaseTrigger | HttpTrigger

export type ApiRoute<T = BaseTrigger> = {
  fn: ApiFunction
  file: string
  functionName: string

  trigger: T
  middleware: HooksMiddleware[]
  functionId: FunctionId
  useInputMetadata?: boolean
}

export function parseApiModule(
  mod: ApiModule,
  file: string,
  router: AbstractRouter
) {
  const apis: ApiRoute[] = []
  for (const [name, fn] of Object.entries(pickBy(mod, isFunction))) {
    const exportDefault = name === 'default'
    const functionName = exportDefault ? EXPORT_DEFAULT_FUNCTION_ALIAS : name
    const functionId = router.getFunctionId(file, functionName, exportDefault)

    const trigger: Trigger = Reflect.getMetadata(OperatorType.Trigger, fn)
    if (!trigger) continue

    // Http Path
    if (trigger.type === HttpTriggerType) {
      trigger.path ??= router.functionToHttpPath(
        file,
        functionName,
        exportDefault
      )
      debug('trigger: %o', trigger)
    }

    // Middleware
    const fnMiddleware = Reflect.getMetadata(OperatorType.Middleware, fn) || []
    const fileMiddleware = mod?.config?.middleware || []
    const middleware = [...fnMiddleware, ...fileMiddleware]

    apis.push({
      fn,
      file,
      functionName,
      functionId,
      trigger,
      middleware,
      useInputMetadata: Reflect.getMetadata(USE_INPUT_METADATA, fn),
    })
  }

  return apis
}
