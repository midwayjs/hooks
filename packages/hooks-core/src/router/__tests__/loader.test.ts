import noop from 'lodash/noop'
import {
  All,
  Api,
  Delete,
  Get,
  Head,
  Middleware,
  Operator,
  OperatorType,
  Options,
  Patch,
  Post,
  Put,
} from '../../api'
import { ApiRouter } from '../api'
import { parseApiModule } from '../loader'

const router = new ApiRouter()

it('load file route with http trigger', () => {
  const routes = parseApiModule(
    {
      get: Api(Get(), async () => {}),
      post: Api(Post(), async () => {}),
      put: Api(Put(), async () => {}),
      del: Api(Delete(), async () => {}),
      patch: Api(Patch(), async () => {}),
      head: Api(Head(), async () => {}),
      options: Api(Options(), async () => {}),
      default: Api(All(), async () => {}),
    },
    '/index.ts',
    router
  )
  expect(routes).toMatchSnapshot()
})

it('load file route with custom trigger', () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      metadata({ setMetadata: defineProperty }) {
        defineProperty(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
        })
      },
    }
  }

  const routes = parseApiModule(
    {
      custom: Api(CustomTrigger(), async () => {}),
    },
    '/index.ts',
    router
  )

  expect(routes).toMatchSnapshot()
})

it('load middleware', () => {
  const routes = parseApiModule(
    {
      get: Api(Get(), Middleware(noop), async () => {}),
      post: Api(Post(), Middleware([noop, noop]), async () => {}),
      config: {
        middleware: [noop],
      },
    },
    '/index.ts',
    router
  )

  expect(routes[0].middleware.length).toBe(2)
  expect(routes[1].middleware.length).toBe(3)
})
