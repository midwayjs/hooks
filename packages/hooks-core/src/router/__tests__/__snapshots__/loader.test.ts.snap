// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load file route with custom trigger 1`] = `
[
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "custom",
    "functionName": "custom",
    "middleware": [],
    "trigger": {
      "isCustom": true,
      "type": "Custom",
    },
    "useInputMetadata": false,
  },
]
`;

exports[`load file route with http trigger 1`] = `
[
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "get",
    "functionName": "get",
    "middleware": [],
    "trigger": {
      "method": "GET",
      "path": "/api/get",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "post",
    "functionName": "post",
    "middleware": [],
    "trigger": {
      "method": "POST",
      "path": "/api/post",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "put",
    "functionName": "put",
    "middleware": [],
    "trigger": {
      "method": "PUT",
      "path": "/api/put",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "del",
    "functionName": "del",
    "middleware": [],
    "trigger": {
      "method": "DELETE",
      "path": "/api/del",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "patch",
    "functionName": "patch",
    "middleware": [],
    "trigger": {
      "method": "PATCH",
      "path": "/api/patch",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "head",
    "functionName": "head",
    "middleware": [],
    "trigger": {
      "method": "HEAD",
      "path": "/api/head",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "options",
    "functionName": "options",
    "middleware": [],
    "trigger": {
      "method": "OPTIONS",
      "path": "/api/options",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
  {
    "file": "/index.ts",
    "fn": [Function],
    "functionId": "index",
    "functionName": "$default",
    "middleware": [],
    "trigger": {
      "method": "ALL",
      "path": "/api/index",
      "requestClient": {
        "client": "@midwayjs/rpc",
        "fetcher": "http",
      },
      "type": "HTTP",
    },
    "useInputMetadata": false,
  },
]
`;
